geomesa-memory/src/main/scala/org/locationtech/geomesa/memory/cqengine/utils/QueryBuilder.scala

```diff
@@ -0,0 +1,196 @@
+/***********************************************************************
+* Copyright (c) 2013-2016 Commonwealth Computer Research, Inc.
+* All rights reserved. This program and the accompanying materials
+* are made available under the terms of the Apache License, Version 2.0
+* which accompanies this distribution and is available at
+* http://www.opensource.org/licenses/apache2.0.php.
+*************************************************************************/
+
+package org.locationtech.geomesa.memory.cqengine.utils
+
+import com.googlecode.cqengine.attribute.Attribute
+import com.googlecode.cqengine.query.simple.{Between, GreaterThan, LessThan, SimpleQuery}
+import org.locationtech.geomesa.filter.PropertyLiteral
+import org.opengis.feature.simple.SimpleFeature
+
+trait CompQueryBuilder {
+  // The type the product queries process, and its class object
+  type VALUE <: Comparable[VALUE]
+  val valueClass: Class[VALUE]    // needed to interact with gt's Java API.
```

Comment:
you might be able to do some fun things using classTags so that you don't have to do so much class matching and not require separate objects for each binding. Of note, you can explicitly pass a classTag using `new ClassTag(binding)`, which is useful in e.g. `addNavigableIndex`.
We did that in the stats package, see:
https://github.com/locationtech/geomesa/blob/master/geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/StatSerializer.scala#L193-L198
https://github.com/locationtech/geomesa/blob/master/geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/EnumerationStat.scala#L21